src/util/amount.rs

```diff
@@ -994,6 +1035,99 @@ pub mod serde {
                 d.deserialize_option(VisitOptAmt::<A>(PhantomData))
             }
         }
+
+        pub mod slice {
+            //! Serialize `&[Amount]` and `&[SignedAmount]` as JSON arrays of numbers denoted in piconero.
+            //! Use with `#[serde(default, serialize_with = "amount::serde::as_pico::slice::serialize")]`.
+
+            use super::super::SerdeAmountForSlice;
+            use serde_crate::{ser::SerializeSeq, Serializer};
+
+            pub fn serialize<A: SerdeAmountForSlice, S: Serializer>(
+                a_slice: &[A],
+                s: S,
+            ) -> Result<S::Ok, S::Error> {
+                let mut seq = s.serialize_seq(Some(a_slice.len()))?;
+
+                for e in a_slice {
+                    e.ser_pico_slice(&mut seq)?;
+                }
+
+                seq.end()
+            }
+        }
+
+        pub mod vec {
+            //! Deserialize a JSON array of numbers (in piconero) into `Vec<Amount>` or
+            //! `Vec<SignedAmount>`.
+            //! It is possible to use `#[serde(default, deserialize_with = "amount::serde::as_pico::vec::deserialize_amount")]`
+            //! for `Vec<Amount>`, and `#[serde(default, deserialize_with = "amount::serde::as_pico::vec::deserialize_signed_amount")]`
+            //! for `Vec<SignedAmount>`.
+
+            use super::super::{Amount, SignedAmount};
+            use core::marker::PhantomData;
+            use serde_crate::{de, Deserializer, __private::size_hint};
+
+            /// Use with `#[serde(default, deserialize_with = "amount::serde::as_pico::vec::deserialize_amount")]`.
+            pub fn deserialize_amount<'d, D: Deserializer<'d>>(
```

Comment:
I tried to do that before and spent a lot of time trying to make it work without trait bounds, but in the end we would need to use trait objects in the return type since more than one type that implement such trait could be returned.

There is also no way to implement deserialize for `[Amount; N]` nor `&[Amount]`

Reply:
Also, when I tried, I implemented `visit_seq` which has a parameter that implements `SeqAccess`, and `SeqAccess` has `next_element` method, which returns `Result<Option<T>, Self::Error>` where `T` needs to implement `Deserialize`, which `Amount` and `SignedAmount` do not implement.

I then tried to make `T` be `u64` or `i64` depending on what `type_prefix()`, returned but didn't work either. I don't remember exactly why, but I think it was related to the parameter that `des_pico` receives implementing `Deserializer`.

Reply:
I'm gonna try again to document the errors here.

Reply:
Thanks for the explanations, all good we keep it as it is now!

Reply:
hmm actually Trait Objects do work since `SerdeAmountForSlice` is `Sized`, which makes total sense.

So the `next_element` error and trying to solve it is the problem.
And just to document, at the top of `visit_seq` we have `let mut vec = Vec::new()`, which must be of type `Vec<X>`. So when making `T` be  u64 or i64 and using `vec.push(Amount::from_pico(amt))`, where amt has type `T`, you get errors like `expected type X, got type Amount`.

Maybe `[std](https://doc.rust-lang.org/std/index.html)::[any](https://doc.rust-lang.org/std/any/index.html)::[Any](https://doc.rust-lang.org/std/any/trait.Any.html#)` could help, but since we have just two types of amount, I don't think it is worth the effort for now.

I also couldn't find any possible other strategy that could make it work generically :/

Reply:
> So the `next_element` error and trying to solve it is the problem.
> And just to document, at the top of `visit_seq` we have `let mut vec = Vec::new()`, which must be of type `Vec<X>`. So when making `T` be u64 or i64 and using `vec.push(Amount::from_pico(amt))`, where amt has type `T`, you get errors like `expected type X, got type Amount`.

I did some quick experiments but I guess it is not worth the effort to find a better solution now, so lets just keep it as it is. I was just wondering if it was possible, looks like you clearly try to make it work and explored!

Reply:
:) ah I mean "trait objects do not work". Forgot the "not".